### ISSUE TITLE
load plugin regression since pip 23.1

### ISSUE THREAD

**Issue by belm0** [pylint-dev/pylint#9208]
### Bug description

```python
Since pip 23.1, plugins can no longer be loaded from local packages other than those declared in setup `packages`.

We had kept our custom plugin modules in a local package that wasn't listed in `packages`, since there is no reason to publish them in the build.  This no longer works.

## create project & env:

echo 'import setuptools; setuptools.setup(name="my_project", packages=["foo"])' > setup.py
mkdir foo bar
touch {foo,bar}/__init__.py

python -m venv .venv
source .venv/bin/activate
pip install --upgrade pip pylint
pip install -e .
```
```


### Configuration

_No response_

### Command used

```shell
pylint --load-plugins bar
```

### Pylint output

```python
************* Module Command line or configuration file
Command line or configuration file:1:0: E0013: Plugin 'bar' is impossible to load, is it installed ? ('No module named 'bar'') (bad-plugin-value)
```

### Expected behavior

plugins from package `bar` can be loaded, even though `bar` is not listed in `packages`

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.8.16 (default, May  9 2023, 17:37:22)
```

### Additional dependencies

```python
pip >= 23.1
```
